### PR TITLE
Add a postprocessor that allows the visualization of the stress.

### LINKED ISSUE #288: implementing a stress post-processor
If anyone has time to do this it would be great to have a post-processor that outputs the stress tensor (sigma_1, sigma_2, sigma_3) and the normal stress.

### PR DESCRIPTION
**Pull request by bangerth**
Also add a test for the strain rate viz postprocessor.

This closes #288.

### CODE REVIEW

**Comment by gassmoeller** [comment]
I agree that apart from @jdannberg 's comments this is a valuable addition to ASPECT's postprocessors for many purposes. However, I am not sure if #288 is closed by this entirely. This postprocessor calculates the shear stress tensor, whereas if I remember correctly @dsarahstamps was looking for the principal components of the full stress tensor. This would mean adding the pressure to the stress tensor as `sigma + pressure * I` and then calculating and outputting the eigenvectors and eigenvalues of this tensor, which are the axis and values of the principal stress directions. @dsarahstamps is this correct? Or can you do all necessary computations in Paraview and are fine with this state?

**Comment by dsarahstamps** [comment]
@gassmoeller You are correct that I am seeking the principal components of the stress tensor in addition to the shear components.

**Comment by ian-r-rose** [comment]
Sarah, I have a function that solves for the eigenvectors and eigenvalues
of a symmetric 3x3/2x2 tensor.  You can find it in this file:

https://github.com/ian-r-rose/aspect/blob/centrifugal/source/postprocess/liouville.cc

If people want to, we could try to add it as a utility during the hackathon.

**Comment by dsarahstamps** [comment]
Thanks Ian. I am interested in working on this contribution to ASPECT at the hackathon.

**Comment by bangerth** [comment]
Let's take an incremental approach: I'm going to change the name of this postprocessor to "shear stress", create another one I call "stress" that also contains the pressure, and let both output all tensor components. At the workshop we can then clone one or the other and create something that outputs the principal components and vectors. Sounds good? I suppose the plan above already gets us 3/4 of the way to where we want to be.

**Comment by gassmoeller** [comment]
Yes, I agree. This postprocessor is still very useful for other purposes. Let me know, when I should have a look through the code.

**Comment by dsarahstamps** [comment]
Yes, that sounds great Wolfgang.

File: include/aspect/postprocess/visualization/shear_stress.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          compute_derived_quantities_vector (const std::vector<Vector<double> >              &uh,
+                                             const std::vector<std::vector<Tensor<1,dim> > > &duh,
+                                             const std::vector<std::vector<Tensor<2,dim> > > &dduh,
+                                             const std::vector<Point<dim> >                  &normals,
+                                             const std::vector<Point<dim> >                  &evaluation_points,
+                                             std::vector<Vector<double> >                    &computed_quantities) const;
+
+          /**
+           * Return the vector of strings describing the names of the computed
+           * quantities. Given the purpose of this class, this is a vector with dim
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/shear_stress.h]: I think the number of entries is now `SymmetricTensor<2,dim>::n_independent_components`

File: include/aspect/postprocess/visualization/shear_stress.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+           * entries all equal to the name given to the constructor.
+           */
+          virtual std::vector<std::string> get_names () const;
+
+          /**
+           * This functions returns information about how the individual components of
+           * output files that consist of more than one data set are to be
+           * interpreted. Since the current class is meant to be used for a single
+           * vector result variable, the returned value is obviously
+           * DataComponentInterpretation::component_is_part repeated dim times.
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/shear_stress.h]: same here

File: include/aspect/postprocess/visualization/stress.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+           * entries all equal to the name given to the constructor.
+           */
+          virtual std::vector<std::string> get_names () const;
+
+          /**
+           * This functions returns information about how the individual components of
+           * output files that consist of more than one data set are to be
+           * interpreted. Since the current class is meant to be used for a single
+           * vector result variable, the returned value is obviously
+           * DataComponentInterpretation::component_is_part repeated dim times.
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/stress.h]: see above

File: include/aspect/postprocess/visualization/stress.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+          compute_derived_quantities_vector (const std::vector<Vector<double> >              &uh,
+                                             const std::vector<std::vector<Tensor<1,dim> > > &duh,
+                                             const std::vector<std::vector<Tensor<2,dim> > > &dduh,
+                                             const std::vector<Point<dim> >                  &normals,
+                                             const std::vector<Point<dim> >                  &evaluation_points,
+                                             std::vector<Vector<double> >                    &computed_quantities) const;
+
+          /**
+           * Return the vector of strings describing the names of the computed
+           * quantities. Given the purpose of this class, this is a vector with dim
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/stress.h]: another one :-)

**Comment by gassmoeller** [comment]
Looks very good. Did you update the test results yet? And the astyle-tester is complaining about a single space. Apart from that I would say it is ready to merge. Thanks for pushing this forward :+1:

**Comment by tjhei** [comment]
Yeah, tests/visualization-stress needs to be corrected.

**Comment by tjhei** [comment]
Also please rebase to current master.

**Comment by bangerth** [comment]
I wasn't actually finished rewriting it all :-) So apologies for the inkomplet test output -- I still needed to add some gravity to the stress test to make sure that there is some pressure. Thanks, though, for all of the very valid corrections!

I think I've addressed it all and rebased. Please take another look.

File: tests/visualization-stress.prm [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# temperature field doesn't matter. set it to zero
+subsection Initial conditions
+  set Model name = function
+  subsection Function
+    set Function expression = x
+  end
+end
+
+
+# no gravity. the pressure will equal just the dynamic component
```
Comment (gassmoeller) [inline tests/visualization-stress.prm]: copy-paste from the other tests, but no longer true in this case ;-)

**Comment by gassmoeller** [comment]
Apart from the one comment, this looks ready to merge. Please fix and merge, whenever you have the time. Thanks for doing this.

**Pull request merged**